File: frontend/src/types/origin.ts
/**
 * Origin flow types — mirror of `backend/app/schemas/origin.py`.
 *
 * Two payloads, one per matrix block ② row sold to Coop Premium. Neither carries
 * an exporter, a destination or a port: those live behind `read:watchai:nominative`
 * and `read:watchai:destinations`, which the tiers reaching these endpoints do not
 * necessarily hold.
 *
 * Note what is nullable and why — the nulls are the interesting part of this API:
 * every one of them means "not measured" rather than "zero", and rendering them
 * as 0 would print a figure that was never observed.
 */

/** The period range a figure actually covers. */
export interface OriginWindow {
  /** `from` is a reserved word in the payload; the backend aliases it. */
  from: string | null;
  to: string | null;
  months: number;
}

export interface MonthlyFlow {
  period: string;
  purchases_t: number;
  exports_beans_t: number;
  exports_transformed_t: number;
  exports_total_t: number;
  /** NULL = STATSER has not published this month yet. It trails the other two
   *  sources by 2-3 months, so the most recent rows are legitimately empty. */
  grinding_declared_t: number | null;
  /** Transformed exports back to bean equivalent (÷ 0.80). NOT the declaration. */
  grinding_derived_t: number;
  /** achats − fèves − broyage déduit, served by the backend so the arithmetic has
   *  one implementation. May be negative for a single month. */
  balance_t: number;
}

export interface YtdComparison {
  source: 'exports' | 'purchases' | 'grinding';
  season: string;
  previous_season: string;
  /** Each source has its OWN window — grinding typically shows 7 months where
   *  exports show 10. Display it: a shared window would imply a collapse that is
   *  purely a publication lag. */
  window: OriginWindow;
  current_t: number;
  previous_t: number;
  /** NULL against a zero baseline — growth off zero is undefined, not infinite. */
  delta_pct: number | null;
}

export interface MonthSynthesis {
  period: string;
  exports_t: number;
  purchases_t: number;
  valcaf_fcfa: number;
  duties_taxes_fcfa: number;
}

export interface ProductMixLine {
  product_code: string;
  /** TRUE for FEVES and HORS_GRADE. Carried so a consumer can see which
   *  denominator a "transformation" figure used: the WatchAI report counts
   *  hors-grade as transformed (27,7 %), Compass counts it as a bean (19,9 %). */
  is_bean_equivalent: boolean;
  export_tonnes: number;
  share_pct: number | null;
}

export interface StatserConfrontation {
  /** Always `gepex` — STATSER only covers those 11 operators, so both sides of
   *  the comparison are computed on them. */
  perimeter: string;
  window: OriginWindow;
  derived_t: number;
  declared_t: number;
  /** derived − declared. Either sign is informative. */
  gap_t: number;
  gap_pct: number | null;
}

export interface TransformationBlock {
  perimeter: string;
  window: OriginWindow;
  purchases_t: number;
  exports_beans_t: number;
  exports_transformed_t: number;
  exports_total_t: number;
  /** Transformed exports converted back to bean equivalent (÷ 0.80). Larger than
   *  the product weight by construction — that is the point, not an error. */
  grinding_derived_t: number;
  balance_t: number;
  balance_pct: number | null;
  /** Derived grinding over PURCHASES. NOT the transformed share of the export
   *  mix — different denominator, ~0.5 pt apart on real data. */
  transformation_rate_pct: number | null;
  outflow_rate_pct: number | null;
  stock_signal: 'stock_constitue' | 'stock_n1_mobilise';
  /** More matter left than was bought. A publishable state, not an error: stock
   *  carries across seasons and the purchase master covers fewer operators than
   *  customs exports. This is why it is a solde *apparent*. */
  outflow_exceeds_purchases: boolean;
  cumulative_balance_t: number;
  monthly_cumulative_t: number[];
  statser_confrontation: StatserConfrontation | null;
}

export interface OriginCampaignResponse {
  data_as_of: string;
  season: string;
  available_seasons: string[];
  perimeter: string;
  monthly: MonthlyFlow[];
  ytd: YtdComparison[];
  month: MonthSynthesis | null;
}

export interface SeasonTotal {
  season: string;
  exports_t: number;
  /** NULL for seasons before the purchase master starts (2020-10). */
  purchases_t: number | null;
}

export interface OriginMarketViewsResponse {
  data_as_of: string;
  season: string;
  available_seasons: string[];
  season_totals: SeasonTotal[];
  monthly: MonthlyFlow[];
  product_mix: ProductMixLine[];
  /** NULL for seasons before the purchase master starts: there are exports and a
   *  product mix, but no balance to compute. Zeros would read as a measurement. */
  transformation: TransformationBlock | null;
}

/** One destination or one port. */
export interface BreakdownLine {
  label: string;
  export_tonnes: number;
  previous_tonnes: number;
  /** Against the SAME months a year earlier, never the previous season in full.
   *  NULL against a zero baseline — growth off zero is undefined. */
  delta_pct: number | null;
  /** Per line, not per view: a destination that stopped shipping in March is
   *  compared over the months it did ship, so two lines can cover different spans. */
  window: OriginWindow;
  share_pct: number | null;
}

export interface DestinationConcentration {
  top1_share_pct: number | null;
  top3_share_pct: number | null;
  count: number;
}

export interface OriginDestinationsResponse {
  data_as_of: string;
  season: string;
  available_seasons: string[];
  previous_season: string;
  destinations: BreakdownLine[];
  ports: BreakdownLine[];
  concentration: DestinationConcentration;
}

/** One named exporter — gated by `read:watchai:nominative`. */
export interface ExporterFlowLine {
  exporter: string;
  is_gepex_member: boolean;
  exports_beans_t: number;
  exports_transformed_t: number;
  exports_total_t: number;
  purchases_t: number;
  grinding_derived_t: number;
  balance_t: number;
  /** This exporter's OWN transformed share (§7). STATSER is a GEPEX aggregate
   *  and is never allocated across operators. */
  transformation_share_pct: number | null;
  previous_exports_t: number;
  /** NULL below the 250 t floor — growth off a tiny base is noise (§8). */
  growth_pct: number | null;
  outflow_exceeds_purchases: boolean;
}

export interface ExporterMover {
  exporter: string;
  growth_pct: number | null;
  exports_total_t: number;
  previous_exports_t: number;
}

export interface OriginExportersResponse {
  data_as_of: string;
  season: string;
  available_seasons: string[];
  previous_season: string;
  growth_floor_tonnes: number;
  exporters: ExporterFlowLine[];
  movers: { up: ExporterMover[]; down: ExporterMover[] };
}

export interface OwnDestinationLine {
  label: string;
  export_tonnes: number;
  share_pct: number | null;
}

export interface BenchmarkPosition {
  exports_total_t: number;
  market_total_t: number;
  market_share_pct: number | null;
  /** Over EVERY exporter, not a truncated top-N. */
  rank: number | null;
  exporters_ranked: number;
  own_destinations: OwnDestinationLine[];
}

export interface OriginBenchmarkResponse {
  data_as_of: string;
  season: string;
  available_seasons: string[];
  previous_season: string;
  /** FALSE is a first-class answer: no exporter identity. Not an empty book —
   *  that would read as "you shipped nothing", which is a different claim. */
  applicable: boolean;
  exporter: string | null;
  position: BenchmarkPosition | null;
}

File: backend/app/schemas/origin.py
"""Response models for the origin flow endpoints (matrix block ②).

Every payload carries ``data_as_of`` (decision #15): ingestion is manual, so there
is no execution log to alert on and staleness is made visible to the user instead
of alerted to ops.

Every *block* additionally carries its own ``window`` and ``perimeter``. That is
not decoration — the three sources stop at different months and the STATSER
confrontation runs on a different population than the balance, so a payload that
omitted them would invite exactly the comparison business-rules §6 forbids.
"""

from __future__ import annotations

from datetime import date
from typing import Optional

from pydantic import BaseModel, Field


class Window(BaseModel):
    """The period range a figure actually covers."""

    from_: Optional[date] = Field(None, alias="from")
    to: Optional[date] = None
    months: int = Field(0, description="Number of months in the window")

    model_config = {"populate_by_name": True}


class MonthlyFlow(BaseModel):
    """One month of the three flows, in tonnes."""

    period: date
    purchases_t: float
    exports_beans_t: float
    exports_transformed_t: float
    exports_total_t: float
    grinding_declared_t: Optional[float] = Field(
        None,
        description="STATSER declaration. NULL means not yet published — it trails "
        "the other two sources by 2-3 months — never zero.",
    )
    grinding_derived_t: float = Field(
        description="Transformed exports converted back to bean equivalent "
        "(/ 0.80). NOT the STATSER declaration above."
    )
    balance_t: float = Field(
        description="achats − exports fèves − broyage déduit, for this month. "
        "Served rather than recomputed client-side so the arithmetic has one "
        "implementation. May be negative for a single month — off-season "
        "shipments draw on stock bought earlier."
    )


class YtdComparison(BaseModel):
    """Season-to-date for ONE source against its own N-1 window (§6).

    Never a shared window across sources: comparing grinding's 7 published months
    against exports' 10 manufactures a collapse that is purely a publication lag.
    """

    source: str = Field(description="exports | purchases | grinding")
    season: str
    previous_season: str
    window: Window
    current_t: float
    previous_t: float
    delta_pct: Optional[float] = Field(
        None, description="NULL against a zero baseline — growth off zero is undefined"
    )


class MonthSynthesis(BaseModel):
    """The four published figures for one month."""

    period: date
    exports_t: float
    purchases_t: float
    valcaf_fcfa: float
    duties_taxes_fcfa: float


class ProductMixLine(BaseModel):
    """One canonical product's share of the season's exports.

    ``is_bean_equivalent`` travels with the line so a consumer can see which
    denominator any "transformation" figure used: the published WatchAI report
    counts HORS GRADE as transformed (27,7 %), Compass counts it as a bean
    (19,9 %). Both are computable from this list; only one is labelled
    transformation.
    """

    product_code: str
    is_bean_equivalent: bool
    export_tonnes: float
    share_pct: Optional[float] = None


class StatserConfrontation(BaseModel):
    """Derived grinding vs STATSER's declaration — a consistency signal (§5).

    Both sides are computed on the GEPEX perimeter. STATSER only covers those 11
    operators, so comparing it to all-country derived grinding would blend a real
    signal with a population mismatch.
    """

    perimeter: str
    window: Window
    derived_t: float = Field(
        description="Transformed exports / 0.80, GEPEX exporters only. Caveat worth "
        "stating in the UI: a group may grind under one legal entity and export "
        "under another, so entity-level filtering can understate this side. The gap "
        "is robust to that — even taking every transformed export in the country as "
        "the numerator it stays negative — but its exact size is not."
    )
    declared_t: float
    gap_t: float = Field(description="derived − declared; either sign is informative")
    gap_pct: Optional[float] = None


class TransformationBlock(BaseModel):
    """Material balance (§4) — bean-equivalent, STATSER not an input."""

    perimeter: str
    window: Window
    purchases_t: float
    exports_beans_t: float
    exports_transformed_t: float
    exports_total_t: float
    grinding_derived_t: float = Field(
        description="Transformed exports converted back to bean equivalent "
        "(/ 0.80). Larger than the product weight by construction."
    )
    balance_t: float
    balance_pct: Optional[float] = None
    transformation_rate_pct: Optional[float] = Field(
        None,
        description="Derived grinding over PURCHASES (§4.3). NOT the transformed "
        "share of the export mix — different denominator, ~0.5 pt apart.",
    )
    outflow_rate_pct: Optional[float] = None
    stock_signal: str
    outflow_exceeds_purchases: bool = Field(
        description="More matter left than was bought over this window. A "
        "publishable state, not an error: stock carries across seasons and the "
        "purchase master covers fewer operators than customs exports (81 vs 102 "
        "on the current batch). This is why it is a solde *apparent*."
    )
    cumulative_balance_t: float
    monthly_cumulative_t: list[float]
    statser_confrontation: Optional[StatserConfrontation] = None


class OriginCampaignResponse(BaseModel):
    """`GET /v1/dashboard/origin/campaign` — the row every tier holds.

    Carries no exporter, destination or port: those are gated by other keys, and
    this endpoint is reachable by tiers that hold neither.
    """

    data_as_of: date
    season: str
    available_seasons: list[str]
    perimeter: str
    monthly: list[MonthlyFlow]
    ytd: list[YtdComparison]
    month: Optional[MonthSynthesis] = None


class OriginMarketViewsResponse(BaseModel):
    """`GET /v1/dashboard/origin/market-views` — aggregated views + transformation."""

    data_as_of: date
    season: str
    available_seasons: list[str]
    season_totals: list["SeasonTotal"]
    monthly: list[MonthlyFlow]
    product_mix: list[ProductMixLine]
    transformation: Optional[TransformationBlock] = Field(
        None,
        description="NULL for seasons before the purchase master starts (2020-10): "
        "there are exports and a product mix, but no balance to compute. Zeros "
        "would read as a measurement.",
    )


class SeasonTotal(BaseModel):
    """Season headline, for the season-comparison view."""

    season: str
    exports_t: float
    purchases_t: Optional[float] = Field(
        None, description="NULL for seasons before the purchase master starts (2020-10)"
    )


OriginMarketViewsResponse.model_rebuild()
